### PR TITLE
Task9 introduce base image loading

### PR DESCRIPTION
**Pull request by kulebin**
Base ImageLoader introduced, no caching so far.

### CODE REVIEW

File: app/src/main/java/com/github/kulebin/myfoursquareapp/adapter/VenueListAdapter.java [hunk 1]
```diff
@@ -64,7 +70,11 @@ public void displayRating(final String rating) {
 
         @Override
         public void displayImage(final String imageUrl) {
-            //todo load and show image
+            imageView.setImageResource(R.drawable.placeholder_foursquare);
+
+            if (imageUrl != null) {
+                IImageLoader.Impl.get().draw(imageUrl, imageView, IMAGE_WIDTH, IMAGE_HEIGHT);
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/adapter/VenueListAdapter.java]: Future improvements: 
- use scaled dimensions, or get image size from imageView
- let placeHolder image as a setting or argument, and use in ImageLoader, not when call loader

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    interface ISizeParams {
+
+        int getWidth();
+
+        int getHeight();
+    }
+
+    private static final int CHUNK_SIZE_POWER = 16; //chunk size = 2^x
+
+    private ISizeParams mSizeParamsCallback;
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: You use this param from different threads!!!!
It's unsafe

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+            @Override
+            public void run() {
+                mSizeParamsCallback = initSizeParamsCallback(pWidth, pHeight);
+
+                IHttpClient.Impl.get().doRequest(pUrl, new IHttpClient.IOnResult<Bitmap>() {
+
+                            @Override
+                            public void onSuccess(final Bitmap pBitmap) {
+                                final Bitmap bitmap = resizeBitmap(pBitmap, pWidth, pHeight);
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: It seems you try to resize image twice :) here, and when you use bitmapOptions.

File: app/src/main/java/com/github/kulebin/myfoursquareapp/adapter/VenueListAdapter.java (hunk 1 above)
Comment (kulebin) [reply app/src/main/java/com/github/kulebin/myfoursquareapp/adapter/VenueListAdapter.java]: Done

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java (hunk 2 above)
Comment (kulebin) [reply app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: Mikhail, this param are initialized and used only from single separated thread

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java (hunk 3 above)
Comment (kulebin) [reply app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: I've changed a little bit logic, but in any case I need resize the bitmap twice, in first place to scale image and second one to crop image to it fits the set width and height

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 4]
```diff
@@ -50,51 +60,112 @@ public Bitmap convert(final InputStream inputStream) throws IOException {
...
+        if (pOptions == null || pOptions.getWidth() == 0 || pOptions.getHeight() == 0) {
+            final DisplayOptions options;
+
+            if (pOptions == null) {
+                options = new DisplayOptions(0);
+            } else {
+                options = pOptions;
+            }
+
+            pView.post(new Runnable() {
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java:87]: Why do we need to run code in such way?

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 5]
```diff
@@ -21,8 +27,12 @@
         int getHeight();
     }
 
+    private static final String TAG = ImageLoader.class.getSimpleName();
     private static final int CHUNK_SIZE_POWER = 16; //chunk size = 2^x
+    private static final int MAX_THREAD_NUMBER = 3;
 
+    private final ExecutorService mExecutorService = Executors.newFixedThreadPool(MAX_THREAD_NUMBER);
+    private final Handler mHandler = new Handler(Looper.getMainLooper());
     private ISizeParams mSizeParamsCallback;
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: It's ok when you load one size images, but can cause you several issues when you should load several big and small images per screen. Let return to this part when needed

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java [hunk 6]
```diff
@@ -50,51 +60,112 @@ public Bitmap convert(final InputStream inputStream) throws IOException {
...
 
                 IHttpClient.Impl.get().doRequest(pUrl, new IHttpClient.IOnResult<Bitmap>() {
 
                             @Override
-                            public void onSuccess(final Bitmap pBitmap) {
-                                final Bitmap bitmap = resizeBitmap(pBitmap, pWidth, pHeight);
-                                
-                                handler.post(new Runnable() {
+                            public void onSuccess(Bitmap pBitmap) {
+                                final Bitmap bitmap = centerCropBitmap(pBitmap, pOptions.getWidth(), pOptions.getHeight());
```
Comment (Mikhail-Ivanou) [inline app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: what if I need a circle bitmap? or real-size? 
Again, lets return to this when really needed

File: app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java (hunk 5 above)
Comment (Mikhail-Ivanou) [reply app/src/main/java/com/github/kulebin/myfoursquareapp/imageLoader/ImageLoader.java]: Probably we can return byte[] from httpClient, or create converter instance  for every call with size

**Pull request merged**